### ISSUE TITLE
[Bug] containers do not resolve other container hostnames

### ISSUE THREAD

**Issue by mr-manuel** [IceWhaleTech/CasaOS#1348]
**Describe the bug**

> A clear and concise description of what the bug is.

It's not possible to resolve the hostname of another container. In a default Docker installation this is possible.

**To Reproduce**

> Steps to reproduce the behavior:

1. Install Grafana
2. Install Node-RED
3. Try to resolve `grafana` on the `node-red` container and/or vice versa, e.g. `docker exec grafana ping node-red`
4. See `127.0.0.1` or `bad address` as result

**Expected behavior**

> A clear and concise description of what you expected to happen.

The IP address of the container should be resolved.

**System Time**

> Run `timedatectl` and share the output

```
               Local time: Wed 2023-08-23 22:01:32 CEST
           Universal time: Wed 2023-08-23 20:01:32 UTC
                 RTC time: Wed 2023-08-23 20:01:32
                Time zone: Europe/Rome (CEST, +0200)
System clock synchronized: yes
              NTP service: active
          RTC in local TZ: no
```

**Additional Information**
Fresh install of two indipendent machines with:
OS: Debian 12
CasaOS Version: v0.4.4.1

```bash
root@casaos:~# docker ps
CONTAINER ID   IMAGE                       COMMAND                  CREATED          STATUS                PORTS                                       NAMES
135f393b0fb3   grafana/grafana:latest      "/run.sh"                34 minutes ago   Up 34 minutes         0.0.0.0:3000->3000/tcp, :::3000->3000/tcp   grafana
6793bb14f1c6   eclipse-mosquitto:2.0.16    "/docker-entrypoint.…"   4 days ago       Up 3 days             0.0.0.0:1883->1883/tcp, :::1883->1883/tcp   mosquitto
cfb703852914   portainer/agent:2.18.4      "./agent"                6 days ago       Up 32 minutes         0.0.0.0:9001->9001/tcp, :::9001->9001/tcp   portainer-agent
2fb39af413ea   influxdb:latest             "/entrypoint.sh infl…"   6 days ago       Up 3 days (healthy)   0.0.0.0:8086->8086/tcp, :::8086->8086/tcp   influxdb
e5764f6c1cf9   nodered/node-red:2.2.2-12   "npm --no-update-not…"   6 days ago       Up 3 days (healthy)                                               node-red
```

**Comment by hyperfelt** [comment]
Same when trying to install Nextcloud and MariaDB, even with both containers in the same network bridge. A workaround for now is to use the IP address of the container you want to reach (you can get it by typing `ip addr` in said container terminal).

**Comment by mr-manuel** [comment]
I red a lot of articles in the last weeks and probably it does not work for the default bridge network, but only for custom created networks in Docker. I found currently no way to change this.

**Comment by jerrykuku** [comment]
CasaOS provides a way around this. After each app is installed, go to settings and in the network options you can see a network with the same name as the current app. Select this bridge network to save and apply, CasaOS will create a new customized bridge network and then will just let other apps use the same bridge network.

**Comment by mr-manuel** [comment]
This sounds not so user friendly and will create 100 networks, if you are installing 100 apps. Thats an overkill for me.

Would it be an idea that all apps use a default bridge network with the name `casaos` or something similar? Then if you want your custom settings like isolation you can change it by your own. My opinion is that out of the box everything should be as user friendly as possible. Therefore I would suggest a default custom bridge network where hostnames are resolving.

**Comment by jerrykuku** [comment]
> This sounds not so user friendly and will create 100 networks, if you are installing 100 apps. Thats an overkill for me.
> 
> Would it be an idea that all apps use a default bridge network with the name `casaos` or something similar? Then if you want your custom settings like isolation you can change it by your own. My opinion is that out of the box everything should be as user friendly as possible. Therefore I would suggest a default custom bridge network where hostnames are resolving.

I agree with your statement. This method is a transitional solution, it doesn't actually create any new bridged networks until you check the new one and save it, the network name you see is just a placeholder.

We'll figure out how to create a default casaos bridge network to solve this problem later.

**Comment by mr-manuel** [comment]
Since the `docker-compose.yml` is already recreated during the install process it should check if there is a network specified in the compose file and is not add the `casaos` network as default. Would that be a solution?

**Comment by jerrykuku** [comment]
Yes, if specify the network in docker-compose.yml, It shouldn't use the default. It's just that none of the apps in our store currently specify a network.
Any way we will discuss a solution as soon as possible.

**Comment by mr-manuel** [comment]
> It's just that none of the apps in our store currently specify a network.

Yes, therefore it should add it automatically when generating the docker-compose.yml. Isn‘t the docker-compose.yml regenerated and saved to another path during installation?

**Comment by jerrykuku** [comment]
We will be scheduling the development of this feature that will automatically bind to a default custom bridge network when the app is installed.

**Comment by Nodiink** [comment]
It doesn't appear to have been written yet: the "Hostname" setting (the last one in the container creation panel) is ignored, and all containers have an auto-generated hostname.
However, it seems that even if you enter the real one, it is not resolved in other containers.

**Comment by KillrOfLife** [comment]
That is stupid behavior that the hostname setting is ignored.
It is default docker behavior, which is used for a lot of homelab configurations to quickly connect containers without exposing them to the rest of the network.

Please let me create a network and add multiple containers to it while keeping hostnames for a simple mysql:3306

**Comment by GCSabas** [comment]
> CasaOS provides a way around this. After each app is installed, go to settings and in the network options you can see a network with the same name as the current app. Select this bridge network to save and apply, CasaOS will create a new customized bridge network and then will just let other apps use the same bridge network.

Quick bit, here.  The "new" network is only created by "Apps" from the store.  Adding a regular container, even ones that are also in the "App Store," will not bring up a new network to bind.

How has something this basic been missing this whole time while these tech YouTubers have just glossed over it?  Did I miss this, somewhere?  Just wild there's no good way to get containers talking to each other, here...

**Comment by infamousjoeg** [comment]
# Workaround until feature is implemented:

## SSH to CasaOS Host
1. `docker network create --label com.docker.compose.network=casaos casaos`

## Log into CasaOS
1. Click the 3 dots in the upper right corner of a tile and select `Settings...`
![image](https://github.com/IceWhaleTech/CasaOS/assets/1924063/143ea3fb-790d-40f8-bfd3-65247caddbe7)
1. Select the `casaos` network you created from the drop-down list in the `Network` section (I named mine `infamousmedia`)
![image](https://github.com/IceWhaleTech/CasaOS/assets/1924063/da38ae71-d0cf-4aaf-9d2e-be7bdac701ac)
1. Set the `Container Hostname` you want to container to use for DNS
![image](https://github.com/IceWhaleTech/CasaOS/assets/1924063/969ef3bc-36a5-4559-a683-bddbd0a20731)
1. Save that and now any app you set to the `casaos` network will be able to communicate with other containers on the same network using the `Container Hostname` set in Step 5.

I've only had these revert back to the default bridge network once on me in several months.  I can't tell if it was when I upgraded the containers or not, so be cautious after you do... you may need to set them back in the `casaos` network after the container is recreated as part of the upgrade process.
